docs/about/The-Engine-architecture.md

```diff
@@ -1,149 +1,324 @@
+kinds of operations that can be performed on this. Blocking the platform task
+runner for an inordinate amount of time may trigger the platform's watchdog, and
+blocking either the UI or raster task runners will cause jank in Flutter
+applications. However, there are tasks necessary for the raster thread that
+require doing some very expensive work. This expensive work is performed on the
+IO task runner.
+
+The main function of the IO task runner is reading compressed images from an
+asset store and making sure these images are ready for rendering on the raster
+task runner. To make sure a texture is ready for rendering, it first has to be
+read as a blob of compressed data (typically PNG, JPEG, etc.) from an asset
+store, decompressed into a GPU friendly format and uploaded to the GPU. These
+operations are expensive and will cause jank if performed on the raster task
+runner. Since only the raster task runner can access the GPU, the IO task runner
+components set up a special context that is in the same sharegroup as the main
+raster task runner context. This happens very early during engine setup and is
+also the reason there is a single task runner for IO tasks. In reality, the
+reading of the compressed bytes and decompression can happen on a thread pool.
+The IO task runner is special because access to the context is only safe from a
+specific thread. The only way to get a resource like
+[`ui.Image`](https://docs.flutter.io/flutter/dart-ui/instantiateImageCodec.html)
+is via an async call; this allows the framework to talk to the IO runner so that
+it can asynchronously perform all the texture operations mentioned. The image
+can then be immediately used in a frame without the raster thread having to do
+expensive work.
+
+There is no way for user code to access this thread either via Dart or native
+plugins. Even the embedder is free to schedule tasks on this thread that are
+fairly expensive. This won’t cause jank in Flutter applications but may delay
+having the futures images and other resources be resolved in a timely manner.
+Even so, it is recommended that custom embedders set up a dedicated thread for
+this task runner.
+
+### Flutter Engine Groups
+
+There are two different ways to create Flutter engines which has ramifications
+on how threading is managed across multiple Flutter engines. Flutter engines can
+be instantiated directly or through the FlutterEngineGroup apis
+[[ios](https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Headers/FlutterEngineGroup.h),
+[android](https://api.flutter.dev/javadoc/io/flutter/embedding/engine/FlutterEngineGroup.html)].
+If an engine is instantiated directly, it gets its own threads. However, all
+engines instantiated with a FlutterEngineGroup will share the same threads
+across that group.  Similarly, engines in a group will share a [dart isolate
+group](https://dart.dev/language/concurrency#performance-and-isolate-groups).
```

Comment:
I think it's worth pointing out that each engine still gets its own root isolate, but now running in the same dart isolate group with all other isolates. And all root isolates from all engines in the group are scheduled to run on the same ui thread(perhaps using the same UI Task Runner?). Spawned worker isolates are running on dart vm thread pool, on the threads that dart vm manages. There is a separate dart vm thread pool for every group.

Reply:
> I think it's worth pointing out that each engine still gets its own root isolate, but now running in the same dart isolate group with all other isolates. And all root isolates from all engines in the group are scheduled to run on the same ui thread(perhaps using the same UI Task Runner?)

Done.  Good point about mentioning the root isolates are unique but share a task runner.  I didn't explain that a dart isolate group implies a thread pool since I think that's better left in the Dart documentation.

Reply:
>  dart isolate group implies a thread pool

Use of single dart isolate group for the whole engine group means shared thread pool for all worker isolates in the engine group, I think it's important to mention that since if affects concurrency characteristics of the multi-engine app.
It basically goes towards something perhaps worth stating explicitly:  that engine group is there to improve resource use(threads, memory, spawn time) and deal with increased dependencies between engines(root isolates run on the same ui thread, dart gc is the same for all engines, worker isolates thread pool is the same for all engines).

Reply:
Okay, I added some of this information, including the shared thread pool.  I think finer details should be in the linked dart documentation.